### DISCUSSION TITLE
OpenTrack and X-Plane 12 in Linux fix

### DISCUSSION THREAD

**Discussion by CannonFodderSE** [opentrack/opentrack#1836] [General]
Greetings
    I'm Running X-Plane 12.0.9-rc-5 on Kubuntu 23.10 Linux with OpenTrack 2.3.0 running in Wine with the freetrack plugin using a USB camera and IR clip.  It all worked at the end of summer/beginning of fall.
    
    After about 6 months of not using it I started it up.  Was unable to get the headtracking to work.  There were several updates to for all programs involved.  The versions above are the latest update I have installed.  Still waiting for the 12.1.0 update for Linux to be released.  After two days of trying every fix or alternate software suggestion it was still broken.
    
    Finally got a working solution by compiling Opentrack natively in Linux, removes Wine overhead, and using amyinorbit's UDP headtracking plugin.
    
    For anybody else having similar issues with headtracking in Linux for X-Plane see below to replicate my solution.

Download headtrack plugin for X-Plane to support head tracking via UDP.
Located on Github at "https://github.com/amyinorbit/headtrack".
Get the latest release.  "https://github.com/amyinorbit/headtrack/releases/tag/2209.1r2" as of this writing.
Download the release as a zip file, "https://github.com/amyinorbit/headtrack/releases/download/2209.1r2/htrack-1010-2209.1r2.zip".
Extract zip file.
Go into the extracted folder labeled "htrack".
Open folder "lin_x64".
Copy  file "htrack.xpl" to your X-Plane plugin folder.  Usually in your X-Plane directory under "Resources/plugins".

Download Opentrack source code from Github and compile using steps below.
These steps were successfully tested on Kubuntu/Ubuntu 23.10.

1 sudo apt install build-essential cmake git qttools5-dev qtbase5-private-dev libprocp2-dev libopencv-dev libproc2-dev
2 git clone https://github.com/opentrack/opentrack
3 cd opentrack
4 cmake . <--don't forget the period
5 make
6 make install
Note: The resulting build output will be placed in the install directory within the opentrack folder created in step 3. It will not "install" itself anywhere outside of the current directory.

Go into the "install" folder followed by the "bin" folder.
Here you will find the Opentrack program labeled "opentrack".
Launch it.
Once opentrack is running, click the pull-down arrow in the "Output" box and select "UDP over network".
Next click the box to the right of the pull-down arrow to configure it.
A new window will popup labeled "UDP protocol settings".
Change the "Remote IP address" to "127" "0" "0" "0".
By default it will computers IP address.
I've found this does not always work.
Using the local loop-back address, 127.0.0.0, always works for me.
Don't change the port.  Leave it at "4242".

Configure the "Options" and "Mapping" to your needs / preferences.

Start X-Plane and launch the aircraft of your choice.
After you are in the cockpit, move your mouse cursor to the top of the screen to open the menu bar.
Click on "Plugins".
You should see "HeadTrack".
Move you mouse to it and click "Track Head Motion".
You should have head tracking working.
There are additional custom settings in the "Setting..." option in the plugin's menu.
Recommend making initial adjustments and response curve settings in Opentrack first then fine tuning in the plugin if needed.

Hope this helps someone else,
Randy Bancroft

**Comment by Lightdasher360** [comment]
Hello! I've been trying to get this working for days now. It seems successful to the point where the tracking registers within the plugin in the sim, but the sim always crashes once I begin to Track Head Motion. Could I possibly ask for help finally fixing the last hurdle?

**Reply by CannonFodderSE** [reply]
Greetings,
Sorry for the late reply.  I was out of town for work.
What version of Linux are you using?
Look at the the "Log.txt" file in your X-Plane folder.  See if there are any errors messages near the end.

**Reply by CannonFodderSE** [reply]
Lightdaher360,
There is a newer version of Opentrack than the version I wrote this for.  I compiled it on my machine and put it [here](https://github.com/CannonFodderSE/opentrack/releases/tag/Opentrack).  Give it a try.

**Reply by Lightdasher360** [reply]
Thank you so so much CannonFodder!! It's finally working for me now! I super duper appreciate your help. :)

And no worries on late reply, sorry for my own lateness! Hope you have a wonderful day!

**Reply by CannonFodderSE** [reply]
Glad it worked for you.  Hope you have many hours of enjoyment.

**Comment by hoover67** [comment]
Thanks very much for this much-needed xplane12 plugin for Linux. I'm using it with opentrack 2026.1.0-19 compiled from source and it worked just fine on the first attempt on Linux Mint 22.3 "zena". 

Mapping headtracking pause & re-center to buttons on my WinWing Orion throttle also appears to work just fine. 

Coffee / beer inbound, thanks again for creating this plugin! 

All the best, 

Uwe

**Comment by talkingerbil** [comment]
Have 2 issues:

What folders/files does opentrack.xpl need to see? I start the sim from a bash script with various env vars set but the plugin doesn't claim UDP port 4242 (checked w/ netstat). It appears to init successfully in sim. I can toggle its status in the plugin manager. If I start the sim straight from the terminal, I do sometimes see the UDP port opened but only sometimes. opentrack itself is not installed at a standard location (set w/ -DCMAKE_INSTALL_PREFIX at build time). Do I need to add something to LD_LIBRARY_PATH or similar evar?

~~Second issue is that OT itself does start tracking and I do see the octopus moving but nothing gets sent to the UDP port (0.0.0.0:4242). Am I missing a config setting?~~ EDIT: OT is sending data. I was accidentally looking at the ethernet device instead of localhost.

**Comment by hoover67** [comment]
Have you selected the correct output plugin in opentrack?

Not at the gaming rig right now so I cannot check the exact name, sorry.

All the best

Uwe

talkingerbil ***@***.***> schrieb am Sa. 11. Juli 2026 um
04:32:

**Comment by talkingerbil** [comment]
If the UDP output plugin isn't the correct one to use, then which one is? I only see 4 options in that list and none of the rest make any sense to my mind.

**Comment by hoover67** [comment]
Yep, that's the one I'm using with 127.0.0.1 as the IP address, Port 4242. Are you using any restrictive appamor settings or firewall rules that might interfere here? 

It's quite weird that port 4242 should only be claimed "sometimes" when the plugin is active in xp12.

**Comment by talkingerbil** [comment]
Rarely, actually. I see the plugin appear in the plugin manager window and there are no errors in Log.txt. Appears to initialize normally. Checking network ports with netstat indicates that no app is claiming port 4242. I put the plugin into Resources/plugins folder and can't see any other configuration options.

I revised my first comment above. OT is indeed sending data to 4242. I was accidentally sniffing on the ethernet device instead of localhost. Not a network routing issue. The plugin should claim the port regardless of any network routing rules. Apparmor isn't installed.

**Comment by talkingerbil** [comment]
Oooo maybe I should RTFM first! Just found this:
https://github.com/opentrack/opentrack/tree/master/x-plane-plugin

**Comment by hoover67** [comment]
Yeah, you and me both it seems. :-) 

I just discovered I'm not even using the opentrack xplane plugin for head tracking, but HeadTrack / htrack by "amyinorbit". More details are available here: 

https://forums.x-plane.org/files/file/67907-headtrack/

No more messing with wine headers etc, just drop the plugin into your xplane plugins folder and use opentrack's udp output plugin. 

works like a charm, it's free but Ami enjoys it if you buy her a coffee :-)

**Comment by talkingerbil** [comment]
I was sending data to 0.0.0.0 which is not the same as 127.0.0.0 for sending. Pity the native opentrack system isn't working.

**Comment by hoover67** [comment]
you meant to write 127.0.0.1, right? :-) 

Anyway, give htrack a try that I linked to above. Works like a charm with opentrack's own udp plugin on Linux, no wine-based workarounds needed. 

All the best, 

Uwe

**Comment by talkingerbil** [comment]
.0 works on my system. It was a surprise I will admit. I randomly had the thought to try 127 instead of 0 and immediately started seeing UDP packets start flowing after changing the first octet, then didn't bother changing the last one to 1. Probably more correct to use a host address as opposed to the network address. I was using htrack for that test.

The wine/proton option is simply referring to the fact that it's using the shared memory-based buffer to write position data (which their native plugin knows how to access) instead of the network stack. The sim isn't running in wine so there's no additional lag or workarounds in play. 

That said, I figured out why the XP option wasn't showing up. I am compiling locally and cmake was finding the wine pkgconfig on my system and defaulting to that option (same general code path). Set **-DSDK_WINE=false** and now it's called XP12 instead of wine/proton in the dropdown menu. Would have to build separate installs of OT to have both.

The htrack option might be better overall though. It has additional smoothing steps that the native OT plugin doesn't offer and works out of the box. The OT plugin is using the raw data stream which is a bit more jumpy in my setup. Maybe I need to play with the camera settings a bit more. Also with htrack, there's no need to fuss with inverting any of the axes or worrying about non-linear mappings (because my camera is offset from the main monitor). I'd need to fiddle with that in the OT gui if I used their plugin.

Only thing I don't like about htrack is that it moves the head position forward to when looking left. I get why...but what if I just want to look out the window?? Might patch/build that locally too.
